modules/core/04-channel/keeper/upgrade.go

```diff
@@ -64,43 +66,165 @@ func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID strin
 
-// upgradeTry
+// ChanUpgradeTry is called by a module to accept the first step of a channel upgrade handshake initiated by
+// a module on another chain. If this function is successful, the proposed upgrade will be returned. If the upgrade fails, the upgrade sequence will still be incremented but an error will be returned.
 func (k Keeper) ChanUpgradeTry(
 	ctx sdk.Context,
 	portID,
 	channelID string,
 	proposedConnectionHops []string,
-	upgradeTimeout types.Timeout,
+	proposedUpgradeTimeout types.Timeout,
 	counterpartyProposedUpgrade types.Upgrade,
 	counterpartyUpgradeSequence uint64,
 	proofCounterpartyChannel,
-	proofCounterpartyUpgrade []byte,
+	proofUpgrade []byte,
 	proofHeight clienttypes.Height,
 ) (types.Upgrade, error) {
-	// TODO
-	return types.Upgrade{}, nil
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	// the channel state must be in OPEN or INITUPGRADE if we are in a crossing hellos situation
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	// verify that the timeout set in UpgradeInit has not passed on this chain
+	if hasPassed, err := counterpartyProposedUpgrade.Timeout.HasPassed(ctx); hasPassed {
+		// abort here and let counterparty timeout the upgrade
+		return types.Upgrade{}, errorsmod.Wrap(err, "upgrade timeout has passed")
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return types.Upgrade{}, errorsmod.Wrap(err, "failed to retrieve connection using the channel connection hops")
+	}
+
+	// make sure connection is OPEN
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return types.Upgrade{}, errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	// assert that the proposed connection hops are compatible with the counterparty connection hops
+	// the proposed connections hops must have a counterparty which matches the counterparty connection hops
+	proposedConnection, err := k.GetConnection(ctx, proposedConnectionHops[0])
+	if err != nil {
+		return types.Upgrade{}, err
+	}
+
+	counterpartyProposedHops := counterpartyProposedUpgrade.Fields.ConnectionHops
+	if proposedConnection.GetCounterparty().GetConnectionID() != counterpartyProposedHops[0] {
+		return types.Upgrade{}, errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnection,
+			"proposed connection hops (%s) does not have counterparty proposed connection hops as counterparty, expected %s, got %s",
+			proposedConnectionHops,
+			counterpartyProposedHops,
+			proposedConnection.GetCounterparty().GetConnectionID(),
+		)
+	}
```

Comment:
https://github.com/cosmos/ibc-go/pull/3687#discussion_r1213202286

Reply:
also see https://github.com/cosmos/ibc-go/pull/3671#discussion_r1211596403

I'd be in favor of making these distinction explicit in the function name:
- upgrade.ValidateBasic
- k.validateSelfUpgrade 
- k.checkUpgradesForCompatibility

Reply:
Cool, rationale sounds good to me! I guess there may be some overlap between validating self upgrade and checking compatibility as both need to do a connection end lookup, right? That's the only thing that would lead me towards having all in one.